server/utils/bot/logger.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package bot
+
+import "fmt"
+
+import "github.com/mattermost/mattermost-plugin-msoffice/server/utils"
+
+type LogContext map[string]interface{}
+
+func level(l string) int {
+	switch l {
+	case "debug":
+		return 4
+	case "info":
+		return 3
+	case "warn":
+		return 2
+	case "error":
+		return 1
+	}
+	return 0
+}
+
+type Logger interface {
+	With(LogContext) Logger
+	Debugf(format string, args ...interface{})
+	Errorf(format string, args ...interface{})
+	Infof(format string, args ...interface{})
+	Warnf(format string, args ...interface{})
+}
+
+func toKeyValuePairs(in map[string]interface{}) (out []interface{}) {
+	for k, v := range in {
+		out = append(out, k)
+		out = append(out, v)
+	}
+	return out
+}
+
+func (bot *bot) With(logContext LogContext) Logger {
```

Comment:
It simplifies writing out kv pairs to the logging messages, right? This is cool, I like it.

Reply:
Yes, it also allows to (a) easily combine it with the printf syntax for the actual messages, and (b) cascade contexts by adding more context to a (local) logger in each function down the stack.